File: frontend/src/lib/expressions/parser.ts
/**
 * Lexer, AST, and parser for pipeline expressions — the TypeScript mirror of
 * `app/pipelines/expressions/parser.py`. Grammar changes must land on both
 * sides and in the shared vectors (`tests/assets/expression_vectors.json`).
 *
 *   expr           := additive
 *   additive       := multiplicative (("+" | "-") multiplicative)*
 *   multiplicative := unary (("*" | "//" | "/" | "%") unary)*
 *   unary          := "-" unary | postfix
 *   postfix        := primary ("." IDENT)*
 *   primary        := INT | FLOAT | STRING | "true" | "false"
 *                   | IDENT | IDENT "(" expr ("," expr)* ")" | "(" expr ")"
 */

import { syntaxError } from "./errors";

export type BinaryOp = "+" | "-" | "*" | "/" | "//" | "%";

export type Expression =
  | { kind: "int"; value: number; position: number }
  | { kind: "float"; value: number; position: number }
  | { kind: "string"; value: string; position: number }
  | { kind: "boolean"; value: boolean; position: number }
  | { kind: "name"; name: string; position: number }
  | { kind: "member"; base: Expression; attribute: string; position: number }
  | { kind: "unary"; operand: Expression; position: number }
  | { kind: "binary"; op: BinaryOp; left: Expression; right: Expression; position: number }
  | { kind: "call"; name: string; args: Expression[]; position: number };

type TokenKind = "int" | "float" | "string" | "ident" | "op" | "eof";

interface Token {
  kind: TokenKind;
  text: string;
  position: number;
}

const OPERATORS = ["//", "+", "-", "*", "/", "%", "(", ")", ",", "."] as const;
const ESCAPES: Record<string, string> = { "\\": "\\", '"': '"', "'": "'", n: "\n", t: "\t" };

const isDigit = (char: string) => char >= "0" && char <= "9";
const isIdentStart = (char: string) => /[A-Za-z_]/.test(char);
const isIdentPart = (char: string) => /[A-Za-z0-9_]/.test(char);

export function tokenize(source: string): Token[] {
  const tokens: Token[] = [];
  let index = 0;
  while (index < source.length) {
    const char = source[index];
    if (/\s/.test(char)) {
      index += 1;
      continue;
    }
    if (isDigit(char)) {
      index = lexNumber(source, index, tokens);
      continue;
    }
    if (isIdentStart(char)) {
      const start = index;
      while (index < source.length && isIdentPart(source[index])) index += 1;
      tokens.push({ kind: "ident", text: source.slice(start, index), position: start });
      continue;
    }
    if (char === '"' || char === "'") {
      index = lexString(source, index, tokens);
      continue;
    }
    const operator = OPERATORS.find((op) => source.startsWith(op, index));
    if (!operator) {
      throw syntaxError(`Unexpected character '${char}'`, index);
    }
    tokens.push({ kind: "op", text: operator, position: index });
    index += operator.length;
  }
  tokens.push({ kind: "eof", text: "", position: source.length });
  return tokens;
}

function lexNumber(source: string, start: number, tokens: Token[]): number {
  let index = start;
  while (index < source.length && isDigit(source[index])) index += 1;
  if (index < source.length && source[index] === ".") {
    const fraction = index + 1;
    if (fraction >= source.length || !isDigit(source[fraction])) {
      throw syntaxError("Expected digits after decimal point", index);
    }
    index = fraction;
    while (index < source.length && isDigit(source[index])) index += 1;
    tokens.push({ kind: "float", text: source.slice(start, index), position: start });
  } else {
    tokens.push({ kind: "int", text: source.slice(start, index), position: start });
  }
  return index;
}

function lexString(source: string, start: number, tokens: Token[]): number {
  const quote = source[start];
  let index = start + 1;
  const parts: string[] = [];
  while (index < source.length) {
    const char = source[index];
    if (char === quote) {
      tokens.push({ kind: "string", text: parts.join(""), position: start });
      return index + 1;
    }
    if (char === "\\") {
      if (index + 1 >= source.length) break;
      const escape = source[index + 1];
      if (!(escape in ESCAPES)) {
        throw syntaxError(`Unknown escape sequence \\${escape}`, index);
      }
      parts.push(ESCAPES[escape]);
      index += 2;
      continue;
    }
    parts.push(char);
    index += 1;
  }
  throw syntaxError("Unterminated string literal", start);
}

const ADDITIVE: readonly string[] = ["+", "-"];
const MULTIPLICATIVE: readonly string[] = ["*", "//", "/", "%"];

class Parser {
  private index = 0;

  constructor(private readonly tokens: Token[]) {}

  private get current(): Token {
    return this.tokens[this.index];
  }

  private advance(): Token {
    const token = this.current;
    this.index += 1;
    return token;
  }

  private matchOp(...operators: string[]): Token | null {
    const token = this.current;
    if (token.kind === "op" && operators.includes(token.text)) {
      return this.advance();
    }
    return null;
  }

  private expectOp(operator: string, context: string): void {
    if (!this.matchOp(operator)) {
      throw syntaxError(`Expected '${operator}' ${context}`, this.current.position);
    }
  }

  parse(): Expression {
    const expr = this.additive();
    if (this.current.kind !== "eof") {
      throw syntaxError(
        `Unexpected '${this.current.text}' after expression`,
        this.current.position,
      );
    }
    return expr;
  }

  private additive(): Expression {
    let left = this.multiplicative();
    let token = this.matchOp(...ADDITIVE);
    while (token) {
      left = {
        kind: "binary",
        op: token.text as BinaryOp,
        left,
        right: this.multiplicative(),
        position: token.position,
      };
      token = this.matchOp(...ADDITIVE);
    }
    return left;
  }

  private multiplicative(): Expression {
    let left = this.unary();
    let token = this.matchOp(...MULTIPLICATIVE);
    while (token) {
      left = {
        kind: "binary",
        op: token.text as BinaryOp,
        left,
        right: this.unary(),
        position: token.position,
      };
      token = this.matchOp(...MULTIPLICATIVE);
    }
    return left;
  }

  private unary(): Expression {
    const token = this.matchOp("-");
    if (token) {
      return { kind: "unary", operand: this.unary(), position: token.position };
    }
    return this.postfix();
  }

  private postfix(): Expression {
    let expr = this.primary();
    let token = this.matchOp(".");
    while (token) {
      const attribute = this.advance();
      if (attribute.kind !== "ident") {
        throw syntaxError("Expected a member name after '.'", token.position);
      }
      expr = { kind: "member", base: expr, attribute: attribute.text, position: token.position };
      token = this.matchOp(".");
    }
    return expr;
  }

  private primary(): Expression {
    const token = this.advance();
    if (token.kind === "int") {
      return { kind: "int", value: Number.parseInt(token.text, 10), position: token.position };
    }
    if (token.kind === "float") {
      return { kind: "float", value: Number.parseFloat(token.text), position: token.position };
    }
    if (token.kind === "string") {
      return { kind: "string", value: token.text, position: token.position };
    }
    if (token.kind === "ident") {
      if (token.text === "true" || token.text === "false") {
        return { kind: "boolean", value: token.text === "true", position: token.position };
      }
      if (this.matchOp("(")) {
        return this.call(token);
      }
      return { kind: "name", name: token.text, position: token.position };
    }
    if (token.kind === "op" && token.text === "(") {
      const expr = this.additive();
      this.expectOp(")", "to close '('");
      return expr;
    }
    if (token.kind === "eof") {
      throw syntaxError("Expression is incomplete", token.position);
    }
    throw syntaxError(`Unexpected '${token.text}'`, token.position);
  }

  private call(name: Token): Expression {
    const args: Expression[] = [];
    if (this.matchOp(")")) {
      return { kind: "call", name: name.text, args, position: name.position };
    }
    args.push(this.additive());
    while (this.matchOp(",")) {
      args.push(this.additive());
    }
    this.expectOp(")", `to close ${name.text}(...)`);
    return { kind: "call", name: name.text, args, position: name.position };
  }
}

/** Parse expression source into an AST; throws a syntax `ExpressionError`. */
export function parse(source: string): Expression {
  if (!source.trim()) {
    throw syntaxError("Expression is empty");
  }
  return new Parser(tokenize(source)).parse();
}

File: app/pipelines/expressions/parser.py
"""Lexer, AST, and parser for pipeline expressions.

One module owns "source text -> AST". The grammar is deliberately small —
arithmetic, a fixed builtin set, string concatenation, and member access on
model variables:

    expr           := additive
    additive       := multiplicative (("+" | "-") multiplicative)*
    multiplicative := unary (("*" | "//" | "/" | "%") unary)*
    unary          := "-" unary | postfix
    postfix        := primary ("." IDENT)*
    primary        := INT | FLOAT | STRING | "true" | "false"
                    | IDENT | IDENT "(" expr ("," expr)* ")" | "(" expr ")"

This grammar is mirrored in TypeScript (`frontend/src/lib/expressions/`) for
live editor feedback; the shared vector suite in
`tests/assets/expression_vectors.json` pins both implementations to identical
behavior. Grammar changes must update both sides and the vectors.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import StrEnum

from app.pipelines.expressions.errors import ExpressionSyntaxError


class TokenKind(StrEnum):
    """Lexical token categories."""

    INT = "int"
    FLOAT = "float"
    STRING = "string"
    IDENT = "ident"
    OP = "op"
    EOF = "eof"


@dataclass(frozen=True)
class Token:
    """One lexical token with its character offset in the source."""

    kind: TokenKind
    text: str
    position: int


_OPERATORS = ("//", "+", "-", "*", "/", "%", "(", ")", ",", ".")
_ESCAPES = {"\\": "\\", '"': '"', "'": "'", "n": "\n", "t": "\t"}
_KEYWORDS = ("true", "false")


def tokenize(source: str) -> list[Token]:
    """Lex the source into tokens, ending with an EOF token."""
    tokens: list[Token] = []
    index = 0
    length = len(source)
    while index < length:
        char = source[index]
        if char.isspace():
            index += 1
            continue
        if char.isdigit():
            index = _lex_number(source, index, tokens)
            continue
        if char.isalpha() or char == "_":
            start = index
            while index < length and (source[index].isalnum() or source[index] == "_"):
                index += 1
            tokens.append(Token(TokenKind.IDENT, source[start:index], start))
            continue
        if char in ('"', "'"):
            index = _lex_string(source, index, tokens)
            continue
        operator = next((op for op in _OPERATORS if source.startswith(op, index)), None)
        if operator is None:
            raise ExpressionSyntaxError(f"Unexpected character {char!r}", index)
        tokens.append(Token(TokenKind.OP, operator, index))
        index += len(operator)
    tokens.append(Token(TokenKind.EOF, "", length))
    return tokens


def _lex_number(source: str, start: int, tokens: list[Token]) -> int:
    """Lex an integer or float literal; return the index after it."""
    index = start
    while index < len(source) and source[index].isdigit():
        index += 1
    if index < len(source) and source[index] == ".":
        fraction = index + 1
        if fraction >= len(source) or not source[fraction].isdigit():
            raise ExpressionSyntaxError("Expected digits after decimal point", index)
        index = fraction
        while index < len(source) and source[index].isdigit():
            index += 1
        tokens.append(Token(TokenKind.FLOAT, source[start:index], start))
    else:
        tokens.append(Token(TokenKind.INT, source[start:index], start))
    return index


def _lex_string(source: str, start: int, tokens: list[Token]) -> int:
    """Lex a quoted string literal; return the index after the closing quote."""
    quote = source[start]
    index = start + 1
    parts: list[str] = []
    while index < len(source):
        char = source[index]
        if char == quote:
            tokens.append(Token(TokenKind.STRING, "".join(parts), start))
            return index + 1
        if char == "\\":
            if index + 1 >= len(source):
                break
            escape = source[index + 1]
            if escape not in _ESCAPES:
                raise ExpressionSyntaxError(f"Unknown escape sequence \\{escape}", index)
            parts.append(_ESCAPES[escape])
            index += 2
            continue
        parts.append(char)
        index += 1
    raise ExpressionSyntaxError("Unterminated string literal", start)


@dataclass(frozen=True)
class Expression:
    """Base AST node; `position` is the character offset in the source."""

    position: int


@dataclass(frozen=True)
class IntLiteral(Expression):
    """Integer literal."""

    value: int


@dataclass(frozen=True)
class NumberLiteral(Expression):
    """Float literal."""

    value: float


@dataclass(frozen=True)
class StringLiteral(Expression):
    """String literal (escapes already resolved)."""

    value: str


@dataclass(frozen=True)
class BooleanLiteral(Expression):
    """`true` / `false` literal."""

    value: bool


@dataclass(frozen=True)
class Name(Expression):
    """Variable reference."""

    name: str


@dataclass(frozen=True)
class Member(Expression):
    """Member access (`base.attribute`) — valid only on model values."""

    base: Expression
    attribute: str


@dataclass(frozen=True)
class Unary(Expression):
    """Unary negation."""

    operand: Expression


@dataclass(frozen=True)
class Binary(Expression):
    """Binary operation: one of + - * / // %."""

    op: str
    left: Expression
    right: Expression


@dataclass(frozen=True)
class Call(Expression):
    """Builtin function call."""

    name: str
    args: tuple[Expression, ...] = field(default_factory=tuple)


_ADDITIVE = ("+", "-")
_MULTIPLICATIVE = ("*", "//", "/", "%")


class _Parser:
    """Recursive-descent parser over the token stream."""

    def __init__(self, tokens: list[Token]) -> None:
        self._tokens = tokens
        self._index = 0

    @property
    def _current(self) -> Token:
        return self._tokens[self._index]

    def _advance(self) -> Token:
        token = self._current
        self._index += 1
        return token

    def _match_op(self, *operators: str) -> Token | None:
        token = self._current
        if token.kind is TokenKind.OP and token.text in operators:
            return self._advance()
        return None

    def _expect_op(self, operator: str, context: str) -> Token:
        token = self._match_op(operator)
        if token is None:
            raise ExpressionSyntaxError(
                f"Expected '{operator}' {context}", self._current.position
            )
        return token

    def parse(self) -> Expression:
        expr = self._additive()
        if self._current.kind is not TokenKind.EOF:
            raise ExpressionSyntaxError(
                f"Unexpected {self._current.text!r} after expression",
                self._current.position,
            )
        return expr

    def _additive(self) -> Expression:
        left = self._multiplicative()
        while (token := self._match_op(*_ADDITIVE)) is not None:
            left = Binary(token.position, token.text, left, self._multiplicative())
        return left

    def _multiplicative(self) -> Expression:
        left = self._unary()
        while (token := self._match_op(*_MULTIPLICATIVE)) is not None:
            left = Binary(token.position, token.text, left, self._unary())
        return left

    def _unary(self) -> Expression:
        if (token := self._match_op("-")) is not None:
            return Unary(token.position, self._unary())
        return self._postfix()

    def _postfix(self) -> Expression:
        expr = self._primary()
        while (token := self._match_op(".")) is not None:
            attribute = self._advance()
            if attribute.kind is not TokenKind.IDENT:
                raise ExpressionSyntaxError(
                    "Expected a member name after '.'", token.position
                )
            expr = Member(token.position, expr, attribute.text)
        return expr

    def _primary(self) -> Expression:
        token = self._advance()
        if token.kind is TokenKind.INT:
            return IntLiteral(token.position, int(token.text))
        if token.kind is TokenKind.FLOAT:
            return NumberLiteral(token.position, float(token.text))
        if token.kind is TokenKind.STRING:
            return StringLiteral(token.position, token.text)
        if token.kind is TokenKind.IDENT:
            if token.text in _KEYWORDS:
                return BooleanLiteral(token.position, token.text == "true")
            if self._match_op("(") is not None:
                return self._call(token)
            return Name(token.position, token.text)
        if token.kind is TokenKind.OP and token.text == "(":
            expr = self._additive()
            self._expect_op(")", "to close '('")
            return expr
        if token.kind is TokenKind.EOF:
            raise ExpressionSyntaxError("Expression is incomplete", token.position)
        raise ExpressionSyntaxError(f"Unexpected {token.text!r}", token.position)

    def _call(self, name: Token) -> Expression:
        args: list[Expression] = []
        if self._match_op(")") is not None:
            return Call(name.position, name.text, tuple(args))
        args.append(self._additive())
        while self._match_op(",") is not None:
            args.append(self._additive())
        self._expect_op(")", f"to close {name.text}(...)")
        return Call(name.position, name.text, tuple(args))


def parse(source: str) -> Expression:
    """Parse expression source text into an AST.

    Raises `ExpressionSyntaxError` (with a character position) on malformed
    input, including empty/blank source.
    """
    if not source.strip():
        raise ExpressionSyntaxError("Expression is empty")
    return _Parser(tokenize(source)).parse()
